detach/registry.go

```diff
@@ -0,0 +1,41 @@
+package detach
+
+import (
+	"context"
+)
+
+type ValueWrap struct {
+	Attach  func(dst, src context.Context) context.Context
+	Cleanup func(ctx context.Context)
+}
+
+var wrapperRegistry []ValueWrap
+
+func RegisterValue(attach func(dst, src context.Context) context.Context) {
+	if attach == nil {
+		panic("cannot register a nil attach function")
+	}
+	RegisterWrap(ValueWrap{Attach: attach})
+}
+
+func RegisterWrap(wrap ValueWrap) {
+	if wrap.Attach == nil {
+		panic("cannot register a ValueWrap with a nil Attach function")
+	}
+	wrapperRegistry = append(wrapperRegistry, wrap)
+}
+
+func attachAll(dst, src context.Context) (context.Context, context.CancelFunc) {
+	var cleanupFuncs []func(ctx context.Context)
+	for _, wrap := range wrapperRegistry {
+		dst = wrap.Attach(dst, src)
+		if wrap.Cleanup != nil {
+			cleanupFuncs = append(cleanupFuncs, wrap.Cleanup)
+		}
+	}
+	return dst, func() {
+		for _, cleanup := range cleanupFuncs {
```

Comment:
we probably need to run cleanup functions FILO, instead of FIFO.

Reply:
that makes sense, although the nature of how you actually register these things kinda means that you can't really rely on ordering.

Reply:
yea good point.

Reply:
alternatively, we could lean into that and shuffle `cleanupFuncs`...